src/circuit/gadgets/constraints.rs

```diff
@@ -483,4 +662,154 @@ mod tests {
+            alloc_bound,
+        );
+        assert!(res.is_ok());
+        assert!(cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_less_than_bound_negative() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let num = s.num(43);
+        let alloc_num =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "num"), s, || Ok(&num)).unwrap();
+        let bound = s.num(42);
+        let alloc_bound =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "bound"), s, || Ok(&bound)).unwrap();
+        let cond = Boolean::Constant(true);
+
+        let res = enforce_less_than_bound(
+            &mut cs.namespace(|| "enforce less than bound"),
+            cond,
+            alloc_num,
+            alloc_bound,
+        );
+        assert!(res.is_ok());
+        assert!(!cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_u64_div_mod() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+
+        let a = s.num(42);
+        let alloc_a = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "a"), s, || Ok(&a)).unwrap();
+        let b = s.num(5);
+        let alloc_b = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "b"), s, || Ok(&b)).unwrap();
+
+        let cond = Boolean::Constant(true);
+
+        let (q, r) = enforce_u64_div_mod(
+            &mut cs.namespace(|| "enforce u64 div mod"),
+            &g,
+            cond,
+            &alloc_a,
+            &alloc_b,
+            s,
+        )
+        .unwrap();
+        assert!(cs.is_satisfied());
+        assert_eq!(q.hash().get_value(), Fr::from_u64(8));
+        assert_eq!(r.hash().get_value(), Fr::from_u64(2));
+    }
+
+    #[test]
+    fn test_enforce_u64_div_mod_zero() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+
+        let a = s.num(42);
+        let alloc_a = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "a"), s, || Ok(&a)).unwrap();
+        let b = s.num(0);
+        let alloc_b = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "b"), s, || Ok(&b)).unwrap();
+
+        let cond = Boolean::Constant(true);
+
+        let (q, r) = enforce_u64_div_mod(
+            &mut cs.namespace(|| "enforce u64 div mod"),
+            &g,
+            cond,
+            &alloc_a,
+            &alloc_b,
+            s,
+        )
+        .unwrap();
+        assert!(cs.is_satisfied());
+        assert_eq!(q.hash().get_value(), Fr::from_u64(0));
+        assert_eq!(r.hash().get_value(), Fr::from_u64(0));
+    }
+
+    #[test]
+    fn test_enforce_boolean_summation() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+
+        let a = s.num(42);
```

Comment:
This is a pretty fragile test, since there's only a single instance being checked. It would be better to check a range/set of values chosen to give higher confidence… Even just N random checks would be better.

Reply:
Now we are testing all numbers from 0 to 128.